packages/components/src/components/SquiggleViewer/ViewerProvider.tsx

```diff
@@ -293,39 +296,66 @@ export function useMergedSettings(path: SqValuePath) {
   return result;
 }
 
-export const ViewerProvider = forwardRef<
-  SquiggleViewerHandle,
-  PropsWithChildren<{
-    partialPlaygroundSettings: PartialPlaygroundSettings;
-    editor?: CodeEditorHandle;
-  }>
->(({ partialPlaygroundSettings, editor, children }, ref) => {
-  const [itemStore] = useState(() => new ItemStore());
-
-  useImperativeHandle(ref, () => ({
-    viewValuePath(path: SqValuePath) {
-      itemStore.scrollToPath(path);
-    },
-  }));
+type Props = PropsWithChildren<{
+  partialPlaygroundSettings: PartialPlaygroundSettings;
+  editor?: CodeEditorHandle;
+}>;
 
-  const [focused, setFocused] = useState<SqValuePath | undefined>();
+export const InnerViewerProvider = forwardRef<SquiggleViewerHandle, Props>(
+  (
+    { partialPlaygroundSettings: unstablePlaygroundSettings, editor, children },
+    ref
+  ) => {
+    const [itemStore] = useState(() => new ItemStore());
+
+    /**
+     * Because we often obtain `partialPlaygroundSettings` with spread syntax, its identity changes on each render, which could
+     * cause extra unnecessary re-renders of widgets, in some cases.
+     * Related discussion: https://github.com/quantified-uncertainty/squiggle/pull/2525#discussion_r1393398447
+     */
+    const playgroundSettings = useStabilizeObjectIdentity(
+      unstablePlaygroundSettings
+    );
+
+    useImperativeHandle(ref, () => ({
+      viewValuePath(path: SqValuePath) {
+        itemStore.scrollToPath(path);
+      },
+    }));
 
-  const globalSettings = useMemo(() => {
-    return merge({}, defaultPlaygroundSettings, partialPlaygroundSettings);
-  }, [partialPlaygroundSettings]);
+    const [focused, setFocused] = useState<SqValuePath | undefined>();
+
+    const globalSettings = useMemo(() => {
+      return merge({}, defaultPlaygroundSettings, playgroundSettings);
+    }, [playgroundSettings]);
+
+    return (
+      <ViewerContext.Provider
+        value={{
+          globalSettings,
+          editor,
+          focused,
+          setFocused,
+          itemStore,
+          initialized: true,
+        }}
+      >
+        {children}
+      </ViewerContext.Provider>
+    );
+  }
+);
+InnerViewerProvider.displayName = "InnerViewerProvider";
+
+export const ViewerProvider = forwardRef<SquiggleViewerHandle, Props>(
+  (props, ref) => {
+    // `ViewerProvider` is a singleton, so if the context already exists, we don't initialize it again
+    const { initialized } = useContext(ViewerContext);
+    if (initialized) {
+      return props.children; // TODO: `ref` and settings will be ignored
```

Comment:
This might create problems in the future.

If another component tries to configure `<ViewerProvider ref={...}>` and then use that ref, but the ref (which is an imperative handle) is ignored then some features (such as "show in editor"; that's the only one that we have now that relies on this ref) won't work correctly.

I believe we don't have any components that would try something like that. Only `<SquigglePlayground>` and `<SquiggleEditor>` use this ref, but they're never wrapped in `<ViewerProvider>`.

I could fix this by exposing `viewValuePath` in context, and then providing another imperative handle here in this component in this branch. But that's too much work for now (I'd have to write `<ProxyViewerProvider>`, since I can't just call `useImperativeHandle` here conditionally...)